doc/Plugin-API.md

```diff
@@ -147,27 +152,30 @@ They each create the proxy to the other side in their constructors by using the
-Utility functions to convert between DTO and API types on the `Ext` side are usually added to [plugin-ext/src/plugin/type-converters.ts](https://github.com/eclipse-theia/theia/blob/master/packages/plugin-ext/src/plugin/type-converters.ts).
-Thus, this is also a good starting point to look for conversion utilities for existing types.
+When [encoding and decoding RPC messages](#encoding-and-decoding-rpc-messages) pure DTO objects can always safely be transmitted.
+For custom classes, we can also adapt the encoding and decoding through extensions.
+However, due to the necessary serialization process, functions and references to other objects can never be transmitted safely.
 
 If functions of objects need to be invoked on the opposite side of their creation, the object needs to be cached on the creation side.
-The other side receives a handle (basically an id) that can be used to invoke the functionality on the creation side.
+The other side receives a handle (usually an id) that can be used to invoke the functionality on the creation side.
 As all cached objects are kept in memory, they should be disposed of when they are no longer needed.
 
 For instance, in [LanguagesExtImpl](https://github.com/eclipse-theia/theia/blob/master/packages/plugin-ext/src/plugin/languages.ts)#registerCodeActionsProvider a new code action provider is created and cached on the `Ext` side and then registered on the `Main` side via its handle.
 When the code action provider’s methods are later invoked on the `Main` side (e.g. in [LanguagesMainImpl](https://github.com/eclipse-theia/theia/blob/master/packages/plugin-ext/src/main/browser/languages-main.ts)#provideCodeActions), it calls the `Ext` side with this handle.
-The `Ext` side then gets the cached object, executes appropriate functions and returns the results back to the `Main` side (e.g. in [LanguageExtImpl](https://github.com/eclipse-theia/theia/blob/master/packages/plugin-ext/src/plugin/languages.ts)#$provideCodeActions).
+The `Ext` side then gets the cached object, executes appropriate functions and returns the results back to the `Main` side (e.g. in [LanguagesExtImpl](https://github.com/eclipse-theia/theia/blob/master/packages/plugin-ext/src/plugin/languages.ts)#$provideCodeActions).
 
 Another example to browse are the [TaskExtImpl](https://github.com/eclipse-theia/theia/blob/master/packages/plugin-ext/src/plugin/tasks/tasks.ts) and [TaskMainImpl](https://github.com/eclipse-theia/theia/blob/master/packages/plugin-ext/src/main/browser/tasks-main.ts) classes.
 
+As we can see from the language and the task example, in Theia we have a complete API protocol that we need provide for plugins that may not be serializable by default.
+In such cases, it has also proven useful to explicitly define a layer of DTO types that are then converted to their API counterpart on the other side.
```

Comment:
I guess this is the part that has me a bit baffled (still), and why I didn't feel comfortable drafting these improvements. Above we mention the use of MsgPackExtensionManager for marshalling custom non-DTO objects, but here we point out this other technique of _explicitly_ defining DTO counterpart types and converting between them. As a reader, I'm wondering...why? Is it historical (predates MsgPackExtensionManager)? Is there some benefit to doing it explicitly this way in some cases? Is it _necessary_ to do it this way in some cases? Which approach should I take?

Reply:
You are right, I believe in theory you could implement all DTOs through the msgpackr extension mechanism and avoid the explicit conversion from domain objects to DTOs and vice versa. I think there are probably several reasons that this is not the case for Theia API:
- msgpackr was only added later and we just added custom serializations for the classes that had that already
- API compliance: VS Code is also using [DTOs](https://github.com/microsoft/vscode/blob/cc7a0865e407e8b5eeaaf0b5cbda6f06f8a99d8b/src/vs/workbench/api/common/extHost.protocol.ts) so it is easier to align with their codebase if we keep the same/a similar system
- msgpackr only allows to register extensions globally so they leak into all connections whether they are related to the custom API or not, see also https://github.com/kriszyp/msgpackr/issues/93. So every single message that is sent may need to be checked against all registered extensions which may impact performance.
- msgpackr only supports [up to 100 extensions](https://github.com/kriszyp/msgpackr?tab=readme-ov-file#custom-extensions) (may not be a big problem for most APIs).

So I think it only makes sense to use msgpackr extensions for very common classes but use DTOs for the actual API. @tsmaeder What do you think about this?

Reply:
It seems good to me...as long as it's in the markdown file and not just this PR comment :-)

I hadn't commented on your response since it looked like you were looking for confirmation from @tsmaeder . If you got that confirmation somehow, then yes...please include that insight in the markdown.

Reply:
Sorry, I kind of missed that one. The main problem with this, IMO, is that we don't know the context in which an object is used in a msgpackr extension. We would have to detect the correct type of object purely from the shape of the object. With 10s or 100s of different object, this seems difficult to maintain correctly whereas if we do the conversion manually, we _know_ the type of object we need to convert and the conversion code is close to its usage. Also, we get instances of interfaces from user code sometimes which we need to send to the front end: these objects may contain extra fields which fool our proposed type detection algorithm into thinking it's the wrong type of object. Or a user might implement an interface with a private field (`_myPrivateField`) and a getter (`myPublicGetter`). You can't detect this from shape of the object if you don't know the type.

Reply:
Thank you for your take on that @tsmaeder! I pushed an update that now should better explain when we use the msgpackr extensions and when we use custom DTOs.